Optimize solid color with single-pixel-buffer-v1

emersion opened the pull request:
References: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/104

~~TODO: bump wayland-protocols version requirement.~~

mstoeckl commented on main.c:
```diff
@@ -114,6 +118,36 @@ static void render_frame(struct swaybg_output *output, cairo_surface_t *surface)
+		uint8_t g8 = (output->config->color >> 16) & 0xFF;
+		uint8_t b8 = (output->config->color >> 8) & 0xFF;
+		uint8_t a8 = (output->config->color >> 0) & 0xFF;
+		double f = (double)0xFFFFFFFF / 0xFF;
+		uint32_t r32 = (uint32_t)(r8 * f);
+		uint32_t g32 = (uint32_t)(g8 * f);
+		uint32_t b32 = (uint32_t)(b8 * f);
+		uint32_t a32 = (uint32_t)(a8 * f);
```
FYI, this section can be simplified a bit.
* `0xFFFFFFFF / 0xFF = 0x01010101` which is an integer, so the casts are unnecessary.
* AFAIK, since background colors are opaque, `a8` will always be `0xff`.

Everything else looks good (although I haven't tested anything.)

emersion replied:
> AFAIK, since background colors are opaque, a8 will always be 0xff.

The man page reads `-c, --color <rrggbb[aa]>`, so I think we need to handle alpha here.

mstoeckl replied:
That line of the man page was updated by https://github.com/swaywm/swaybg/pull/45 .

emersion replied:
Ah, makes sense!

That said, I think I'd prefer to keep the current code even if it's unused:

- The current cairo codepath supports it.
- It would actually be more work and inconsistencies to assert that the alpha channel is 0xFF, and then hardcode the single pixel buffer param to 0xFFFFFFFF.

mstoeckl replied:
Yes, that is reasonable.

mstoeckl commented on main.c:
```diff
@@ -126,6 +160,7 @@ static void render_frame(struct swaybg_output *output, cairo_surface_t *surface)
 	cairo_paint(cairo);
 	cairo_restore(cairo);
 	if (output->config->mode == BACKGROUND_MODE_SOLID_COLOR) {
+		// TODO: use a shm single-pixel buffer and scale it with viewporter if available
```
This comment can now be removed..

mstoeckl commented:
When I ran this branch of swaybg on an output with 2x scaling, `swaybg -c '#ff00ff'` failed to display a background; see following partial log with WAYLAND_DEBUG=1.

```
[  76026.226] zwlr_layer_surface_v1@11.configure(355, 1280, 720)
[  76026.242]  -> zwlr_layer_surface_v1@11.ack_configure(355)
[  76026.258]  -> wp_single_pixel_buffer_manager_v1@8.create_u32_rgba_buffer(new id wl_buffer@10, 4294967295, 0, 4294967295, 4294967295)
[  76026.275]  -> wl_surface@3.set_buffer_scale(2)
[  76026.289]  -> wl_surface@3.attach(wl_buffer@10, 0, 0)
[  76026.304]  -> wl_surface@3.damage_buffer(0, 0, 2147483647, 2147483647)
[  76026.319]  -> wp_viewporter@7.get_viewport(new id wp_viewport@12, wl_surface@3)
[  76026.333]  -> wp_viewport@12.set_destination(2560, 1440)
[  76026.346]  -> wl_surface@3.commit()
[  76026.357]  -> wp_viewport@12.destroy()
[  76026.368]  -> wl_buffer@10.destroy()
[  76027.542] wl_display@1.error(wl_surface@3, 2, "Buffer size (1x1) is not divisible by scale (2)")
wl_surface@3: error 2: Buffer size (1x1) is not divisible by scale (2)
```

mstoeckl commented:
Also, a minor issue, not necessarily worth fixing: every time the output is resized, swaybg creates a new pair of `wl_buffer` and `wl_viewport` objects, and then promptly destroys them after use; it would be slightly more efficient to create one buffer/viewport pair on first use, and then call only ` wp_viewport.set_destination(...)` and `wl_surface.commit(...)` when a size update is needed.

emersion commented:
> so the following diff may be sufficient to fix the issue.

Indeed! After discussing with pq in #wayland, it seems like it's a good practice to not use `wl_surface.set_buffer_scale` together with viewporter.

I've addressed all of the comments I think, thanks for the feedback!

emersion commented:
Thanks for the review!
